### ISSUE TITLE
ICE "data-layout... differs from LLVM default"

### ISSUE THREAD

**Issue by PeterGV** [Trantect/win_driver_example#5]
My primary interest is in Windows kernel-mode drivers, and I've just recently started playing with Rust.

In trying to build (either target for) this project using "rustc 1.47.0-nightly (770bd3d1d 2020-08-10)" I get:

`error: internal compiler error: src\librustc_codegen_llvm\context.rs:159:13: data-layout for builtin `x86_64-pc-windows-msvc` target, `e-m:w-i64:64-f80:128-n8:16:32:64-S128`, differs from LLVM default, `e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128``

and the compiler eventually panics.

Because I'm a complete Rust noob, I have no insight into this whatsoever... even though I've tried to work my way through target definitions and the "data_layout" specification.

Any help would be appreciated.

**Comment by harryfei** [comment]
We use `rustc +nightly -Z unstable-options --print target-spec-json --target x86_64-unknown-linux-gnu` to generate the `x86_64-kernel-windows-msvc.json` (with fixing some necessary fields). I think you can replace `data-layout`'s value with the expected one `e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128`.

**Comment by PeterGV** [comment]
Thanks for the quick answer.  I definitely appreciate it.

That, plus the hack to define **_fltused** (from the other open issue) has gotten the build to succeed for me.

I'm pressed for time on a separate project right now, so it will take me a few days to be able to test this.  But I thank you very much and will report back here with my test results.

**Comment by PeterGV** [comment]
I apologize for the delay.  I did want to circle-back and thank you again for your help.

I have managed to build the driver successfully, and it loads and runs under Win10 20H1.  

Thank you VERY much for your help on this.  I'm hoping to get some time to see if I can expand on your work here.  There's a lot of interest, and maybe some significant value, in facilitating Windows driver development in Rust right now.  There don't seem to be a lot of people who are both Windows driver devs, and Rust devs.  I'm hoping we can change that a bit.

Anyhow... many thanks for your help.
